applications/FluidDynamicsApplication/custom_constitutive/newtonian_two_fluid_2d_law.cpp

```diff
@@ -0,0 +1,136 @@
+//************************************************************************************
+
+NewtonianTwoFluid2DLaw::NewtonianTwoFluid2DLaw(const NewtonianTwoFluid2DLaw& rOther)
+    : Newtonian2DLaw(rOther)
+{}
+
+//********************************CLONE***********************************************
+//************************************************************************************
+
+ConstitutiveLaw::Pointer NewtonianTwoFluid2DLaw::Clone() const {
+    return Kratos::make_shared<NewtonianTwoFluid2DLaw>(*this);
+}
+
+//*******************************DESTRUCTOR*******************************************
+//************************************************************************************
+
+NewtonianTwoFluid2DLaw::~NewtonianTwoFluid2DLaw() {}
+
+std::string NewtonianTwoFluid2DLaw::Info() const {
+    return "NewtonianTwoFluid2DLaw";
+}
+
+double NewtonianTwoFluid2DLaw::ComputeEffectiveViscosity(ConstitutiveLaw::Parameters& rParameters) const 
+{
+    double viscosity;
+    EvaluateInPoint(viscosity, DYNAMIC_VISCOSITY, rParameters);
+    const Properties& prop = rParameters.GetMaterialProperties();
+
+    if (prop.Has(C_SMAGORINSKY)) {
+        const double csmag = prop[C_SMAGORINSKY];
+        if (csmag > 0.0) {
+            double density;
+            EvaluateInPoint(density, DENSITY, rParameters);
+            const double strain_rate = EquivalentStrainRate(rParameters);
+            const BoundedMatrix<double,3,2>& rDN_DX = rParameters.GetShapeFunctionsDerivatives();
+            const double elem_size = ElementSizeCalculator<2,3>::GradientsElementSize(rDN_DX);
+            double length_scale = std::pow(csmag * elem_size, 2);
+            viscosity += 2.0*length_scale * strain_rate * density;
+        }
+    }
+    return viscosity;
+}
+
+void NewtonianTwoFluid2DLaw::EvaluateInPoint(
+    double& rResult,
+    const Variable<double>& rVariable,
+    ConstitutiveLaw::Parameters& rParameters) const
+{
+    const SizeType n_nodes = 3;
+    const GeometryType& r_geom = rParameters.GetElementGeometry();
+    const array_1d<double,n_nodes>& rN = rParameters.GetShapeFunctionsValues();
+
+    // Compute Gauss pt. distance value
+    double dist = 0.0;
+    for (unsigned int i = 0; i < n_nodes; ++i){
+        dist += rN[i] * r_geom[i].FastGetSolutionStepValue(DISTANCE);
+    }
+
+    // Compute the Gauss pt. value considering the previously compute distance value
+    rResult = 0.0; // Initialize result variable value
+    SizeType n_avg = 0; // Number of nodes on the same side as the gauss point
+    for (unsigned int i = 0; i < n_nodes; ++i) {
+        if ( dist * r_geom[i].FastGetSolutionStepValue(DISTANCE) > 0.0) {
+            n_avg += 1;
+            rResult += r_geom[i].FastGetSolutionStepValue(rVariable);
+        }
+    }
+    rResult /= n_avg;
+}
+
+double NewtonianTwoFluid2DLaw::EquivalentStrainRate(ConstitutiveLaw::Parameters& rParameters) const
+{
+    // Calculate Symetric gradient (Voigt notation)
+    const SizeType n_nodes = 3;
```

Comment:
should this be hard-coded?

Reply:
@swenczowski we already have an utility that does this kind of operations for you. Please check this file and try to use the methods there whenever possible (we can also add new ones if you see they are missing):
https://github.com/KratosMultiphysics/Kratos/blob/master/applications/FluidDynamicsApplication/custom_utilities/fluid_element_utilities.h

Reply:
I had another close look and I am very sorry, but I cannot explain how the files newtonian_two_fluid_2d_law.cpp and newtonian_two_fluid_2d_law.h were added to my git push. I apologize for a possible mistake that may have been caused due to my lack of experience concerning Git(Hub).

@GuillermoCasas : I am sorry for asking, but do you happen to know where these files come from?

PS: The only (really small) changes I made to the constitutive law were directly implemented in "custom_utilities/two_fluid_navier_stokes_data.h" (CalculateAirMaterialResponse).

Reply:
@swenczowski No... But did you not start your work based on a pre-existing branch? Maybe those changes come from that original branch and had never been pushed to the master. Would that not make sense?

Reply:
@swenczowski I think that I implemented it when creating the Python solver (actually what @GuillermoCasas comments). I think that at those days we did not have the utilities @jcotela is pointing out. Would you mind to take the chance to include them?

Reply:
@rubenzorrilla I will have a look and try my best. Thank you for the clarification of the origin of the files.

Reply:
The new push uses functions from custom_utilities/two_fluid_navier_stokes_data.h. Unfortunately, in my implementation, this does not make the code much shorter or nicer. I hope that I did not miss one perfectly fitting function for this purpose...

Reply:
Thinking some more about it, I think that the strain rate is an input parameter for the constitutive law, so you should not even need to compute it (just take its norm). We usually get it as
```C++
    const Vector& r_strain_rate = rParameters.GetStrainVector(); 
```
(which is confusing because the method names of the constitutive law assume a solid, but that's another story). Can you check that you have something there in your case? The element should be providing this value. It does for the base fluid_element class, but I'm not sure if we are overwriting it in the two-fluid version (that's why I'd ask you to check).

Reply:
The function .GetStrainVector does not seem to be overwritten by FluidElement or any class derived from FluidElement. Thus,  
`const Vector& S = rParameters.GetStrainVector();`
was introduced and the relevant unit test were run without problems.